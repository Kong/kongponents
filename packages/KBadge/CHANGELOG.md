# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.1.9](https://github.com/Kong/kongponents/compare/@kongponents/kbadge@0.1.8...@kongponents/kbadge@0.1.9) (2020-03-06)

**Note:** Version bump only for package @kongponents/kbadge





## [0.1.8](https://github.com/Kong/kongponents/compare/@kongponents/kbadge@0.1.6...@kongponents/kbadge@0.1.8) (2020-01-16)

**Note:** Version bump only for package @kongponents/kbadge





## [0.1.6](https://github.com/Kong/kongponents/compare/@kongponents/kbadge@0.1.5...@kongponents/kbadge@0.1.6) (2019-11-12)

**Note:** Version bump only for package @kongponents/kbadge





## [0.1.5](https://github.com/Kong/kongponents/compare/@kongponents/kbadge@0.1.4...@kongponents/kbadge@0.1.5) (2019-11-11)

**Note:** Version bump only for package @kongponents/kbadge





## [0.1.4](https://github.com/Kong/kongponents/compare/@kongponents/kbadge@0.1.3...@kongponents/kbadge@0.1.4) (2019-11-08)


### Bug Fixes

* **dist:** dist to all packages ([#152](https://github.com/Kong/kongponents/issues/152)) ([217c46b](https://github.com/Kong/kongponents/commit/217c46b02afe960f8fa597ba77410dcf034b1716))





## [0.1.3](https://github.com/Kong/kongponents/compare/@kongponents/kbadge@0.1.2...@kongponents/kbadge@0.1.3) (2019-11-08)

**Note:** Version bump only for package @kongponents/kbadge





# @kongponents/kbadge

## Table of Contents

- [0.1.2](#012)
- [0.1.1](#011)

### 0.1.2
- Update appearance (success, warning, danger) background colors for accessible contrasts. [#122](https://github.com/Kong/kongponents/pull/122)

### 0.1.1
- :rocket: initialÂ release, [#110](https://github.com/Kong/kongponents/pull/110)
