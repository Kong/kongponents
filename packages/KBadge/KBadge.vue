<template>
  <div
    :class="[appearance !== 'custom' && `kbadge-${appearance}`]"
    :style="color && backgroundColor && {backgroundColor, color}"
    class="k-badge"
  >
    <slot />
  </div>
</template>

<script>
export const appearances = {
  success: 'success',
  danger: 'danger',
  warning: 'warning'
}

export default {
  props: {
    /**
      * Base styling<br>
      * One of [danger, warning, success etc.]
      */
    appearance: {
      type: String,
      required: false,
      validator: function (value) {
        return Object.keys({...appearances, 'custom': ''}).indexOf(value) !== -1
      },
      default: 'custom'
    },

    color: {
      type: String,
      required: false,
      default: ''
    },

    backgroundColor: {
      type: String,
      required: false,
      default: ''
    }
  }
}
</script>

<style lang="scss" scoped>
@import '~@kongponents/styles/_variables.scss';

.k-badge {
  display: inline-flex;
  justify-content: center;
  font-weight: 500;
  font-size: var(--KBadgeFontSize, var(--type-sm, type(sm)));
  height: 27px;
  align-items: center;
  width: var(--KBadgeWidth, 71px);
  padding: var(--KBadgePaddingY, 2px) var(--KBadgePaddingX, 4px);
  font-family: var(--font-family-sans, font(sans));
  border-radius: var(--KBadgeBorderRadius, 3px);

  // Appearances
  &.kbadge-success {
    color: var(--KBadgeSuccessColor, var(--green-dark, color(green-dark)));
    border-color: var(--KBadgeSuccessBorder, var(--green-lighter, color(green-lighter)));
    background-color: var(--KBadgeSuccessBackground, var(--green-lighter, color(green-lighter)));
  }
  &.kbadge-danger {
    color: var(--KBadgeDangerColor, var(--red-dark, color(red-dark)));
    border-color: var(--KBadgeDangerBorder, var(--red-lighter, color(red-lighter)));
    background-color: var(--KBadgeDangerBackground, var(--red-lighter, color(red-lighter)));
  }
  &.kbadge-warning {
    color: var(--KBadgeWarningColor, var(--yellow-darker, color(yellow-darker)));
    border-color: var(--KBadgeWarningBorder, var(--yellow-lighter, color(yellow-lighter)));
    background-color: var(--KBadgeWarningBackground, var(--yellow-lighter, color(yellow-lighter)));
  }
}
</style>
