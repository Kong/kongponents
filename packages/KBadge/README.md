# @kongponents/kbadge

[![](https://img.shields.io/npm/v/@kongponents/kbadge.svg?style=flat-square)](https://www.npmjs.com/package/@kongponents/kbadge)

```vue
<KBadge appearance="success">SUCCESS</KBadge>
<KBadge appearance="warning">WARNING</KBadge>
<KBadge appearance="danger">DANGER</KBadge>
<KBadge color="var(--blue-lighter)" background-color="var(--blue-base)">👋🏻 Hello</KBadge>
```
