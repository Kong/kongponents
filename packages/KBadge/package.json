{
  "name": "@kongponents/kbadge",
  "version": "6.27.1",
  "description": "Badges, pills, lozenges, or whatever you wanna call them.",
  "main": "dist/KBadge.umd.min.js",
  "componentName": "KBadge",
  "source": "KBadge.vue",
  "files": [
    "dist",
    "KBadge.vue"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Kong/kongponents.git"
  },
  "author": "Kong Inc",
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/Kong/kongponents/issues"
  },
  "homepage": "https://github.com/Kong/kongponents#readme",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@kongponents/styles": "^6.27.1"
  }
}
