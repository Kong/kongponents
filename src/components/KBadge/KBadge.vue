<template>
  <div
    class="k-badge"
    :class="[appearance, { 'method': isMethodBadge }]"
  >
    <KTooltip :text="showTooltip ? tooltip : undefined">
      <div
        class="badge-content"
        :class="{ 'icon-after': !iconBefore }"
        :tabindex="showTooltip ? 0 : undefined"
      >
        <slot
          v-if="$slots.icon"
          name="icon"
        />
        <div
          ref="badgeTextElement"
          class="badge-content-wrapper"
        >
          <slot />
        </div>
      </div>
    </KTooltip>
  </div>
</template>

<script lang="ts">
import type { PropType } from 'vue'
import { ref, computed, nextTick, onMounted } from 'vue'
import KButton from '@/components/KButton/KButton.vue'
import KTooltip from '@/components/KTooltip/KTooltip.vue'
import type { BadgeAppearance } from '@/types'
import { BadgeAppearances, BadgeMethodAppearances } from '@/types'
import useUtilities from '@/composables/useUtilities'

const { getSizeFromString } = useUtilities()

// Must explicitly define components so KTooltip works in tests
export default {
  name: 'KBadge',
  components: { KButton, KTooltip },
}
</script>

<script setup lang="ts">
const props = defineProps({
  appearance: {
    type: String as PropType<BadgeAppearance>,
    required: false,
    default: 'info',
    validator: (value: string): boolean => {
      return Object.keys(BadgeAppearances).includes(value)
    },
  },
  /**
   * Tooltip text that will be displayed on hover.
   */
  tooltip: {
    type: String,
    default: '',
  },
  /**
   * Whether tooltip should only be shown when the badge is truncated.
   */
  truncationTooltip: {
    type: Boolean,
    default: false,
  },
  /**
   * Max width to apply truncation at
   * Is superseded by CSS variable if both provided
   */
  maxWidth: {
    type: String,
    default: '200',
  },
  /**
   * A boolean whether or not to show the icon before the badge text (or after).
   */
  iconBefore: {
    type: Boolean,
    default: true,
  },
})

const isMethodBadge = computed(() => {
  return Object.keys(BadgeMethodAppearances).includes(props.appearance)
})

const badgeTextElement = ref<HTMLDivElement | null>()

const isTruncated = ref<boolean>(false)

const maxWidth = computed((): string => getSizeFromString(props.maxWidth))

const setTruncation = async (): Promise<void> => {
  if (badgeTextElement.value) {
    await nextTick()
    isTruncated.value = badgeTextElement.value?.offsetWidth < badgeTextElement.value?.scrollWidth
  }
}

const showTooltip = computed((): boolean => {
  if (!props.tooltip) {
    return false
  }

  if (props.truncationTooltip) {
    return isTruncated.value
  }

  return true
})

onMounted(() => {
  setTruncation()
})
</script>

<style lang="scss" scoped>
/* Component variables */

$kBadgeMethodWidth: 85px;

/* Component styles */

.k-badge {
  // apply info appearance by default (in case of invalid appearance)
  @include badgeAppearance;

  // fixing mixed-decls deprecation: https://sass-lang.com/d/mixed-decls
  // stylelint-disable-next-line no-duplicate-selectors
  & {
    @include badgeWrapper;
  }

  > div:not(.k-popover) {
    display: flex;
  }

  .badge-content {
    @include badgeContent;

    &.icon-after {
      flex-direction: row-reverse;
    }
  }

  .badge-content-wrapper {
    @include truncate;

    max-width: v-bind('maxWidth');
  }

  :deep(#{$kongponentsKongIconSelector}) {
    height: var(--kui-icon-size-30, $kui-icon-size-30) !important;
    width: var(--kui-icon-size-30, $kui-icon-size-30) !important;
  }

  :deep([role="button"]) {
    &:not([disabled]) {
      cursor: pointer;
    }

    // adopts info appearance hover styles by default (in case of invalid appearance)

    &[disabled] {
      color: var(--kui-color-text-disabled, $kui-color-text-disabled) !important;
      pointer-events: none;
    }
  }

  &.method {
    .badge-content {
      justify-content: center;
      min-width: $kBadgeMethodWidth !important;
      text-align: center;
      text-transform: uppercase;
    }
  }

  /* Appearances */

  &.info {
    @include badgeAppearance;
  }

  &.success {
    @include badgeAppearance(var(--kui-color-background-success-weakest, $kui-color-background-success-weakest), var(--kui-color-text-success, $kui-color-text-success), var(--kui-color-text-success-strong, $kui-color-text-success-strong));
  }

  &.warning {
    @include badgeAppearance(var(--kui-color-background-warning-weakest, $kui-color-background-warning-weakest), var(--kui-color-text-warning, $kui-color-text-warning), var(--kui-color-text-warning-strong, $kui-color-text-warning-strong));
  }

  &.danger {
    @include badgeAppearance(var(--kui-color-background-danger-weakest, $kui-color-background-danger-weakest), var(--kui-color-text-danger, $kui-color-text-danger), var(--kui-color-text-danger-strong, $kui-color-text-danger-strong));
  }

  &.decorative {
    @include decorativeBadgeAppearance;
  }

  &.neutral {
    @include badgeAppearance(var(--kui-color-background-neutral-weaker, $kui-color-background-neutral-weaker), var(--kui-color-text-neutral-strong, $kui-color-text-neutral-strong), var(--kui-color-text-neutral-stronger, $kui-color-text-neutral-stronger));
  }

  // methods
  &.connect {
    @include badgeAppearance(var(--kui-method-color-background-connect, $kui-method-color-background-connect), var(--kui-method-color-text-connect, $kui-method-color-text-connect), var(--kui-method-color-text-connect-strong, $kui-method-color-text-connect-strong));
  }

  &.custom {
    @include badgeAppearance(var(--kui-color-background-neutral-weak, $kui-color-background-neutral-weak), var(--kui-color-text, $kui-color-text), var(--kui-color-text-neutral-stronger, $kui-color-text-neutral-stronger));
  }

  &.delete {
    @include badgeAppearance(var(--kui-method-color-background-delete, $kui-method-color-background-delete), var(--kui-method-color-text-delete, $kui-method-color-text-delete), var(--kui-method-color-text-delete-strong, $kui-method-color-text-delete-strong));
  }

  &.get {
    @include badgeAppearance(var(--kui-method-color-background-get, $kui-method-color-background-get), var(--kui-method-color-text-get, $kui-method-color-text-get), var(--kui-method-color-text-get-strong, $kui-method-color-text-get-strong));
  }

  &.head {
    @include badgeAppearance(var(--kui-method-color-background-head, $kui-method-color-background-head), var(--kui-method-color-text-head, $kui-method-color-text-head), var(--kui-method-color-text-head-strong, $kui-method-color-text-head-strong));
  }

  &.options {
    @include badgeAppearance(var(--kui-method-color-background-options, $kui-method-color-background-options), var(--kui-method-color-text-options, $kui-method-color-text-options), var(--kui-method-color-text-options-strong, $kui-method-color-text-options-strong));
  }

  &.patch {
    @include badgeAppearance(var(--kui-method-color-background-patch, $kui-method-color-background-patch), var(--kui-method-color-text-patch, $kui-method-color-text-patch), var(--kui-method-color-text-patch-strong, $kui-method-color-text-patch-strong));
  }

  &.post {
    @include badgeAppearance(var(--kui-method-color-background-post, $kui-method-color-background-post), var(--kui-method-color-text-post, $kui-method-color-text-post), var(--kui-method-color-text-post-strong, $kui-method-color-text-post-strong));
  }

  &.put {
    @include badgeAppearance(var(--kui-method-color-background-put, $kui-method-color-background-put), var(--kui-method-color-text-put, $kui-method-color-text-put), var(--kui-method-color-text-put-strong, $kui-method-color-text-put-strong));
  }

  &.trace {
    @include badgeAppearance(var(--kui-method-color-background-trace, $kui-method-color-background-trace), var(--kui-method-color-text-trace, $kui-method-color-text-trace), var(--kui-method-color-text-trace-strong, $kui-method-color-text-trace-strong));
  }
}
</style>
