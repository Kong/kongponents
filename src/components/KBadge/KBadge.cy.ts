import KBadge from '@/components/KBadge/KBadge.vue'
import type { BadgeAppearance, BadgeSize } from '@/types'
import { BadgeAppearances, BadgeSizes } from '@/types'

const rendersCorrectAppearance = (variant: BadgeAppearance) => {
  it(`renders KBadge with the ${variant} appearance`, () => {
    cy.mount(KBadge, {
      props: {
        appearance: variant,
      },
      slots: {
        default: () => variant,
      },
    })

    cy.get('.k-badge').should('have.class', variant)
  })
}

const rendersCorrectSize = (size: BadgeSize) => {
  it(`sets ${size} class when size passed`, () => {
    cy.mount(KBadge, {
      props: {
        size,
      },
      slots: {
        default: () => size.charAt(0).toUpperCase() + size.substring(1).toLowerCase(),
      },
    })

    cy.get('.k-badge').should('have.class', size)
  })
}

describe('KBadge', () => {
  // Loop through BadgeAppearances
  Object.keys(BadgeAppearances).map(a => rendersCorrectAppearance(a as BadgeAppearance))

  // Loop through BadgeSizes
  Object.values(BadgeSizes).map(s => rendersCorrectSize(s))

  it('defaults to info `appearance`', () => {
    cy.mount(KBadge, {
      slots: {
        default: () => 'Hello!',
      },
    })

    cy.get('.k-badge').should('have.class', 'info')
  })

  it('displays `tooltip` at all times', () => {
    const tooltipText = 'Hello! Long badge with truncated text here'

    cy.mount(KBadge, {
      props: {
        tooltip: tooltipText,
      },
      slots: {
        default: () => 'Hello!',
      },
    })

    cy.get('.k-tooltip').should('contain.text', tooltipText)
  })

  it('when `truncationText` is true, only displays `tooltip` if truncated', () => {
    const tooltipText = 'Hello!'

    cy.mount(KBadge, {
      props: {
        tooltip: tooltipText,
        truncationTooltip: true,
      },
      slots: {
        default: () => 'Hello!',
      },
    })

    cy.get('.k-tooltip').should('not.exist')
  })

  it('it should apply `maxWidth` prop when provided', () => {
    const maxWidth = '10px'
    cy.mount(KBadge, {
      props: {
        maxWidth,
      },
      slots: {
        default: () => 'Hello!',
      },
    })

    cy.get('.badge-content').should('have.css', 'max-width').and('eq', maxWidth)
  })

  it('renders the icon slot', () => {
    cy.mount(KBadge, {
      slots: {
        icon: '<span data-testid="icon">Icon</span>',
      },
    })

    cy.get('[data-testid="icon"]').should('exist').should('be.visible')
  })
})
