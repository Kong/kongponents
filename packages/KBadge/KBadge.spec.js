import { mount } from '@vue/test-utils'
import { default as KBadge, appearances } from '@/KBadge/KBadge'

const rendersCorrectAppearance = (variant) => {
  it(`renders KBadge with the ${variant} appearance`, () => {
    const wrapper = mount(KBadge, {
      propsData: {
        'appearance': variant
      },
      slots: {
        default: variant
      }
    })

    expect(wrapper.find('.k-badge').classes()).toContain(`kbadge-${variant}`)
    expect(wrapper.html()).toMatchSnapshot()
  })
}

describe('KBadge', () => {
  Object.keys(appearances).map(a => rendersCorrectAppearance(a))

  it('matches snapshot', () => {
    const wrapper = mount(KBadge)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('defaults to default badge', () => {
    const wrapper = mount(KBadge, {
      slots: {
        default: 'Hello!'
      }
    })

    expect(wrapper.find('.k-badge-default')).toBeDefined()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('handles custom colors', () => {
    const wrapper = mount(KBadge, {
      propsData: {
        appearance: 'custom',
        color: 'white',
        backgroundColor: 'red'
      },
      slots: {
        default: 'Hello!'
      }
    })

    expect(wrapper.find('.k-badge').element.style.color).toEqual('white')
    expect(wrapper.find('.k-badge').element.style.backgroundColor).toEqual('red')
    expect(wrapper.html()).toMatchSnapshot()
  })
})
