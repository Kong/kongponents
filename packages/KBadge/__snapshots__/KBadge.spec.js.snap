// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KBadge defaults to default badge 1`] = `<div class="k-badge truncate kbadge-default">Hello!</div>`;

exports[`KBadge handles custom colors 1`] = `<div class="k-badge truncate" style="background-color: red; color: white;">Hello!</div>`;

exports[`KBadge matches snapshot 1`] = `<div class="k-badge truncate kbadge-default"></div>`;

exports[`KBadge renders KBadge with the danger appearance 1`] = `<div class="k-badge truncate kbadge-danger">danger</div>`;

exports[`KBadge renders KBadge with the default appearance 1`] = `<div class="k-badge truncate kbadge-default">default</div>`;

exports[`KBadge renders KBadge with the success appearance 1`] = `<div class="k-badge truncate kbadge-success">success</div>`;

exports[`KBadge renders KBadge with the warning appearance 1`] = `<div class="k-badge truncate kbadge-warning">warning</div>`;
